### ISSUE TITLE
RwLock::try_read is unsound

### ISSUE THREAD

**Issue by zakarumych** [mvdnes/spin-rs#123]
This safe code creates two references to the save value, one of which is mutable.

```
let lock = RwLock::new(1);
let read = lock.read();
for _ in 0 .. usize::MAX >> 2 {
  forget(lock.read());
}
let mut write = lock.write();

(&*read, &mut *write)
```

It will succeed, due to overflow in internal atomic value.

There must be protection against overflows in locking functions.
For reference, see https://docs.rs/atomicell/0.1.2/src/atomicell/borrow.rs.html#95

**Comment by zesterer** [comment]
Hi, thanks very much for reporting this. Any chance you could take a glance at the fix here? https://github.com/mvdnes/spin-rs/pull/124

**Comment by zesterer** [comment]
'Fixed' by #124
